gc/structs/ForwardedHeader.hpp

```diff
@@ -95,6 +95,12 @@ class MM_ForwardedHeader
 	static const uintptr_t _copySizeAlignement = (uintptr_t)SIZE_ALIGNMENT;
 	static const uintptr_t _minIncrement = (131072 & _remainingSizeMask); /**< min size of copy section; does not have to be a power of 2, but it has to be aligned with _copySizeAlignement */ 
 #endif /* OMR_GC_CONCURRENT_SCAVENGER */
+#if defined(OMR_GC_VLHGC)
+	/* the grow tag is used by VLHGC and can only be set if the FORWARDED_TAG is already set.  It signifies that the object grew a hash field when moving */
+	static const uintptr_t _grow_tag = (uintptr_t)0x2;
```

Comment:
There is a potential we can get rid of Grow flag and related APIs.... What do you think is preventing us? I don't see it being used. 

I'm aware that GROW flag matches COPY_HINT for CS (value of 0x2) used for large copy optimization. Those are both flags that are set in forwarded header (in source object). GROW cannot exist at the same bit position if we want the same optimization eventually be used in concurrent Copy-Forward.

If GROW bit is indeed used, than one of GROW or  COPY_HINT should move to 0x8





1:47
if GROW bit is indeed used, than one of GROW or  COPY_HINT should move to 0x8

Reply:
Correction: in forwarded header there are only 3 lower bits available (it contains non-compressed destination object address that is 8 byte aligned), so bit 0x8 is not available.

Reply:
My concern here was regarding `MM_ForwardedHeader::getForwardedObject()`. If we're in a platform that does not support concurrent scavenger (meaning `OMR_GC_CONCURRENT_SCAVENGER` is not defined) but it does support VLHGC we would have something like:
```
omrobjectptr_t
MM_ForwardedHeader::getForwardedObject()
{
	omrobjectptr_t forwardedObject = NULL;
	uintptr_t forwardedTag = _forwardedTag;
	if (isForwardedPointer()) {
		forwardedObject = (omrobjectptr_t)(getCanonicalPreserved() & ~forwardedTag);
	}

	return forwardedObject;
}
```
Not sure if this scenario is possible but this would result in a different `forwardedTag` than `_all_tags` which is the combination of grow and forward tag. And grow tag is also used in `setForwardedObjectGrowing()`. 
Also isn't GROW tag and BEING_COPIED_HINT tag mutually exclusive, meaning one is solely used by scavenger and the other for VLHGC? If that's the case could they become one tag?

Reply:
That bit-wise-and operation in getForwardedObject strips out any flags (within 3 LSBs as mentioned earlier) that could be set in forwarded header (via setForwardedObject(Growing), to get the pure forwarded address that is properly 8 byte aligned.  Forwarded bit is for sure there and must be removed. Grow flag needs to be removed only if it was set (along with forwarded address). But I see no reason to set that flag to start with, if no-one relies on it. Indeed, didObjectGrowOnCopy is not used, and there are no references of GROW flag other than set/getForwardedObject APIs.

BEING_COPIED_HINT  is currently exclusive for Concurrent Scavenger, but at some point we might need BEING_COPIED_HINT flag in VLHGC as well, so we need to remove GROW flag.

So, let's try just using plain set/getForwardedObject, instead.

Reply:
I was hasty, you're right, there's no reason to keep GROW flag around, hence no reason to keep ALL_TAG flag either. The only usage of `setForwardedObjectGrowing()` is in `runtime/gc_vlhgc/CopyForwardScheme.cpp` which can be safely replaced by `setForwardedObject()`. With that we don't need `setForwardedObjectGrowing()`, `didObjectGrowOnCopy()` nor `getForwardedObjectVLHGC()`. I removed all those and replaced them with the equivalent `MM_ForwardedHeader` calls and all tests have passed. This means that we don't need this PR at all.